pkg/ingester/ingester_v2.go

```diff
@@ -1096,13 +1095,44 @@ func (i *Ingester) closeAllTSDB() {
 // concurrently opening TSDB.
 func (i *Ingester) openExistingTSDB(ctx context.Context) error {
 	level.Info(util.Logger).Log("msg", "opening existing TSDBs")
-	wg := &sync.WaitGroup{}
-	openGate := gate.New(i.cfg.BlocksStorageConfig.TSDB.MaxTSDBOpeningConcurrencyOnStartup)
 
-	// Keep track of all errors that could occur.
+	wg := &sync.WaitGroup{}
+	queue := make(chan string)
 	errs := tsdb_errors.MultiError{}
 	errsMx := sync.Mutex{}
 
+	// Create a pool of workers which will open existing TSDBs.
+	for n := 0; n < i.cfg.BlocksStorageConfig.TSDB.MaxTSDBOpeningConcurrencyOnStartup; n++ {
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+
+			for userID := range queue {
+				startTime := time.Now()
+
+				db, err := i.createTSDB(userID)
+				if err != nil {
+					errsMx.Lock()
+					errs.Add(errors.Wrapf(err, "unable to open TSDB for user %s", userID))
+					errsMx.Unlock()
+
+					level.Error(util.Logger).Log("msg", "unable to open TSDB", "err", err, "user", userID)
+					return
```

Comment:
By returning here, we lose a consumer of `queue` channel, which can lead to blocking of dir walker, when it tries to send another item to the channel.

Reply:
Uh good point!